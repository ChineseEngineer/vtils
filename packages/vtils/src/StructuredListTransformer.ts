import {isPlainObject} from './is'

export type StructuredList<TItem = any> = TItem[]

export interface PackedStructuredList<TItem = any> {
  readonly __IS_PACKED_STRUCTURED_LIST__: true,
  readonly keys: Array<keyof TItem>,
  readonly values: Array<Array<TItem[keyof TItem]>>,
}

/**
 * 结构化列表数据转换器。
 */
export class StructuredListTransformer<TItem> {
  constructor(private list: StructuredList<TItem> | PackedStructuredList<TItem>) {}

  /**
   * 如果是打包后的结构化列表数据，则解包后返回，否则直接返回。
   *
   * @param value 数据
   * @returns 返回结果数据
   */
  static transformIfNeeded(value: any) {
    if (isPlainObject(value) && (value as PackedStructuredList).__IS_PACKED_STRUCTURED_LIST__ === true) {
      return new StructuredListTransformer(value as any).unpack()
    }
    return value
  }

  /**
   * 打包结构化列表数据。
   *
   * @returns 返回打包后的结构化列表数据
   */
  pack(): PackedStructuredList<TItem> {
    const structuredList: StructuredList<TItem> = this.list as any
    const keys: Array<keyof TItem> = Object.keys(structuredList[0]) as any
    const values = []
    for (const item of structuredList) {
      values.push(
        keys.map(key => item[key]),
      )
    }
    return {
      __IS_PACKED_STRUCTURED_LIST__: true,
      keys: keys,
      values: values,
    }
  }

  /**
   * 解包结构化列表数据。
   *
   * @returns 返回解包后的结构化列表数据
   */
  unpack(): StructuredList<TItem> {
    const packedStructuredList: PackedStructuredList<TItem> = this.list as any
    const structuredList: TItem[] = []
    for (const values of packedStructuredList.values) {
      const item: TItem = {} as any
      for (let i = 0, len = values.length; i < len; i++) {
        item[packedStructuredList.keys[i]] = values[i]
      }
      structuredList.push(item)
    }
    return structuredList
  }
}
