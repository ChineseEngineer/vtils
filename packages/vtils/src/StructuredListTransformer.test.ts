import {PackedStructuredList, StructuredList, StructuredListTransformer} from './StructuredListTransformer'

test('表现正常', () => {
  interface Item {
    id: number,
    name: string,
  }

  const structuredList: StructuredList<Item> = [
    {id: 1, name: 'Jay'},
    {id: 2, name: 'Jay2'},
    {id: 3, name: 'Jay3'},
    {id: 4, name: 'Jay4'},
  ]
  const packedStructuredList: PackedStructuredList<Item> = new StructuredListTransformer(structuredList).pack()
  const unpackedStructuredList: StructuredList<Item> = new StructuredListTransformer(packedStructuredList).unpack()

  expect(packedStructuredList).toMatchSnapshot('packedStructuredList')
  expect(unpackedStructuredList).toEqual(structuredList)
  expect(StructuredListTransformer.transformIfNeeded(packedStructuredList)).toEqual(structuredList)
  expect(StructuredListTransformer.transformIfNeeded(1)).toEqual(1)
  expect(StructuredListTransformer.transformIfNeeded([])).toEqual([])
})
