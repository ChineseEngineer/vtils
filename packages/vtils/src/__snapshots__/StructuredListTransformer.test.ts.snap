// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`表现正常: pack 1`] = `
Object {
  "__IS_PACKED_STRUCTURED_LIST__": true,
  "keys": Array [
    "id",
    "name",
  ],
  "values": Array [
    Array [
      1,
      "Jay",
    ],
    Array [
      2,
      "Jay2",
    ],
    Array [
      3,
      "Jay3",
    ],
    Array [
      4,
      "Jay4",
    ],
  ],
}
`;

exports[`表现正常: packedStructuredList 1`] = `
Object {
  "__IS_PACKED_STRUCTURED_LIST__": true,
  "keys": Array [
    "id",
    "name",
  ],
  "values": Array [
    Array [
      1,
      "Jay",
    ],
    Array [
      2,
      "Jay2",
    ],
    Array [
      3,
      "Jay3",
    ],
    Array [
      4,
      "Jay4",
    ],
  ],
}
`;
